src/util/parallel/data.ml

```diff
@@ -0,0 +1,242 @@
+open Batteries
+
+module type DefaultType = sig
+  type t
+  val default: unit -> t
+  val to_string: t -> string
+end
+
+module ConcurrentBucket (Key: Hashtbl.HashedType) (Val: DefaultType) = struct
+  type t = {
+    key: Key.t;
+    value: Val.t Atomic.t;
+    next: t option Atomic.t;
+  }
+
+  let create key = {
+    key = key;
+    value = Atomic.make @@ Val.default ();
+    next = Atomic.make None;
+  }
+
+  let create_with_value key value = {
+    key = key;
+    value = value;
+    next = Atomic.make None;
+  }
+
+  let create_with_value_and_next key value next = {
+    key = key;
+    value = value;
+    next = Atomic.make (Some next);
+  }
+
+  let find_option sll key =
+    let rec find sll key =
+      if Key.equal sll.key key then Some sll.value
+      else match Atomic.get sll.next with
+        | None -> None
+        | Some next -> find next key
+    in
+    find sll key
+
+  let find sll key =
+    match find_option sll key with
+    | None -> failwith "Key not found"
+    | Some value -> value
```

Comment:
Why not just `Option.get`? Is having our own `failwith` better than that raising?

Reply:
Having a more specific error message is good, because it can save the user of this module a dive in to (even though a very brief one). Here we could go with both, but in the outwards facing parts I have a strong preference for an error message that describes what went wrong on a level that is relevant to the user.

Reply:
`raise Not_found` might be the best option. It's what the standard library raises.

Reply:
Although, if there's always a default value (https://github.com/goblint/analyzer/pull/1748#discussion_r2098696583), then why should this ever fail?

Reply:
The default value is created on query (similar to pythons `defaultdict`, there are cases where we do not want this.

Using `find` instead of `find_create` in places where we expect the value to be present makes it easier to catch bugs.

Any ideas on how to make the behaviour and the rationale more obvious? At the very least it would make sense to document this in the mli I guess.